src/org/ods/orchestration/util/MROPipelineUtil.groovy

```diff
@@ -296,90 +299,74 @@ class MROPipelineUtil extends PipelineUtil {
-                        } else if (this.project.isAssembleMode && PipelinePhases.FINALIZE) {
-                            new FinalizeOdsComponent(project, steps, git, logger).run(repo, baseDir)
-                        } else {
-                            this.logger.debug("Repo '${repo.id}' is of type ODS Service Component. Nothing to do in phase '${name}' for target environment '${targetEnvToken}'.")
-                        }
-                    } else if (repo.type?.toLowerCase() == PipelineConfig.REPO_TYPE_ODS_TEST) {
-                        if (name == PipelinePhases.TEST) {
-                            executeODSComponent(repo, baseDir)
-                        } else {
-                            this.logger.debug("Repo '${repo.id}' is of type ODS Test Component. Nothing to do in phase '${name}' for target environment '${targetEnvToken}'.")
-                        }
-                    } else {
-                        def phaseConfig = repo.pipelineConfig.phases ? repo.pipelineConfig.phases[name] : null
-                        if (phaseConfig) {
-                            def label = "${repo.id} (${repo.url})"
-
-                            if (phaseConfig.type == PipelineConfig.PHASE_EXECUTOR_TYPE_MAKEFILE) {
-                                this.steps.dir(baseDir) {
-                                    def steps = "make ${phaseConfig.target}"
-                                    this.steps.sh script: steps, label: label
-                                }
-                            } else if (phaseConfig.type == PipelineConfig.PHASE_EXECUTOR_TYPE_SHELLSCRIPT) {
-                                this.steps.dir(baseDir) {
-                                    def steps = "./scripts/${phaseConfig.steps}"
-                                    this.steps.sh script: steps, label: label
+    String logRepoSkip(String phase, Map repo) {
+        def targetEnvToken = this.project.buildParams.targetEnvironmentToken
+        this.logger.info("Repo '${repo.id}' is of type '${repo.type}'. Nothing to do in phase '${phase}' for target environment '${targetEnvToken}'.")
+    }
+
+    void executeRepoGroups(List<Set<Map>> repoGroups, Closure repoExecute, Closure preRepoExecute = null, Closure postRepoExecute = null) {
+        repoGroups.each { repoGroup ->
+            def reposInGroup = repoGroup.collect { it.id }
+            this.logger.debug("Executing repo group consisting of: ${reposInGroup}")
+            Map<String, Closure> parallelTasks = repoGroup.collectEntries { repo ->
+                [
+                    repo.id,
+                    {
+                        this.executeBlockAndFailBuild {
+                            if (preRepoExecute) {
+                                preRepoExecute(repo)
+                            }
+                            if (repoExecute) {
+                                this.steps.dir(repoBaseDir(repo)) {
```

Comment:
This is where `steps.dir(baseDir)` is now centrally applied, allowing it to be removed from all other places.